nft_api/src/ingest/main.rs

```diff
@@ -45,217 +59,488 @@ struct AppSpecificRev {
+
+                            // Get owner from index 4.
+                            //let owner = keys.index(instruction.accounts[4] as usize);
+                            let owner = keys.get(instruction.accounts[4] as usize);
+                            let owner = String::from_utf8(owner.key().unwrap().to_vec()).unwrap();
+
+                            // Get new owner from index 5.
+                            //let new_owner = keys.index(instruction.accounts[5] as usize);
+                            let new_owner = keys.get(instruction.accounts[5] as usize);
+                            let new_owner =
+                                String::from_utf8(new_owner.key().unwrap().to_vec()).unwrap();
+
+                            // Get message and leaf.
+                            let hex_message = hex::encode(&add.message);
+                            let leaf =
+                                keccak::hashv(&[&new_owner.as_bytes(), add.message.as_slice()]);
+
+                            // Populate app event.
+                            app_event.op = String::from("tran");
+                            app_event.tree_id = tree_id;
+                            app_event.leaf = leaf.to_string();
+                            app_event.message = hex_message;
+                            app_event.owner = owner;
+                            app_event.new_owner = Some(new_owner);
                         }
-                        Err(e) => {
-                            eprintln!("{}", e.to_string())
+                        gummyroll_crud::InstructionName::Remove => {
+                            // Get data.
+                            let data = instruction.data[8..].to_owned();
+                            let data_buf = &mut data.as_slice();
+                            let remove: gummyroll_crud::instruction::Remove =
+                                gummyroll_crud::instruction::Remove::deserialize(data_buf).unwrap();
+
+                            // Get tree ID.
+                            //let tree_id = keys.index(instruction.accounts[3] as usize);
+                            let tree_id = keys.get(instruction.accounts[3] as usize);
+                            let tree_id =
+                                String::from_utf8(tree_id.key().unwrap().to_vec()).unwrap();
+
+                            // Get owner from index 0.
+                            //let owner = keys.index(instruction.accounts[0] as usize);
+                            let owner = keys.get(instruction.accounts[0] as usize);
+                            let owner = String::from_utf8(owner.key().unwrap().to_vec()).unwrap();
+
+                            // Get leaf.
+                            let leaf = bs58::encode(&remove.leaf_hash).into_string();
+
+                            // Populate app event.
+                            app_event.op = String::from("rm");
```

Comment:
can we just set `app_event.op = gummyroll_crud::get_instruction_type(&instruction.data)`? I think this would make `app_event_to_databse` cleaner

Reply:
This is a good idea, I would like to do this in a follow on PR given that it may allow me to do more significant refactoring to the code (plus this PR is already pretty big).